neps/nep-0371.md

```diff
@@ -0,0 +1,149 @@
+---
+NEP: 371
+Title: Runtime Function Call Access Key Info
+Author: Ben Kurrek <ben.kurrek@near.org>, Jakob Meier <jakob@near.org>
+DiscussionsTo: https://gov.near.org/t/expose-gas-price-and-contract-access-key-info-within-runtime/24788
+Status: Draft
+Type: Standards Track
+Category: Review
+Created: 13-Jul-2022
+---
+
+## Summary
+
+There is currently no way to query for information about access keys stored on a contract at runtime. Adding the host function `access_key_info` to the runtime specification will allow to query this for the current account an easy way. This fully covers the most important use cases that are currently known for querying access keys at runtime.
+
+## Motivation
+
+As contracts start to use access keys in more creative ways, developers will want to know information such as the access key's allowance, receiver and more. Since contracts are often built for scalability, Gas and storage costs are often delegated to the end user. If you charge the user for an access key and its allowance, when the key is deleted or used up, that allowance should be refunded to the purchaser. Currently, there is no way to query for the leftover allowance before a key is deleted. This information can tell you how much Gas was spent, and allows for a slew of different scenarios to arise.
+
+## Rationale and alternatives
+
+Without creating a promise, we can only query for information stored on the same shard. This is why in this proposal, the access key information at runtime will only come from keys stored on the current contract. We had also investigated the possibility of introducing pagination for all keys stored on the contract but this is infeasible at the moment due to the significant redesign requirement of the VM logic.
+
+## Runtime specification
+
+The contract runtime provides one host function called `access_key_info` which is callable in regular calls and views.
+
+```rust
+fn access_key_info (
```

Comment:
So this API can only be used to access info about a known key?  What if a contract wants to know the actual set of keys associated with it?  Maybe there is already an API for that or is that not an interesting use case?

Reply:
It could interesting to paginate keys, we just haven't had anyone coming up with an actual use case, yet. :)

And because the API for that would be significantly trickier to implement,  we opted it out, for now.

I expect most calls to either look up the current access key (`signer_account_pk()` already exists for that) or otherwise take an argument from outside that specifies which access keys to check. Listing all keys through rpc REST API has been supported for a long time, so the signer side should always be able to provide the list if necessary.

Reply:
>And because the API for that would be significantly trickier to implement

To expand on this, we don't *really* have a way to iterate keys in our runtime.